### DISCUSSION TITLE
how to get human readable output like strace?

### DISCUSSION THREAD

**Discussion by minhuw** [bpftrace/bpftrace#3088] [Q&A]
I want to know what socket options are set and get in my system. So I write a simple bpftrace script:

```
#!/usr/bin/env bpftrace

tracepoint:syscalls:sys_enter_setsockopt
{
    printf("%s [%d] set %d-%d\n", comm, pid, args.level, args.optname);
}

tracepoint:syscalls:sys_enter_getsockopt
{
    printf("%s [%d] get %d-%d\n", comm, pid, args.level, args.optname);
}
```

It generally works well. But it prints raw data difficult to read:
```
systemd-timesyn [848] get 1-7
systemd-timesyn [848] set 1-7
systemd-timesyn [848] get 1-7
systemd-timesyn [848] set 1-32
systemd-timesyn [848] get 1-7
systemd-timesyn [848] set 1-7
systemd-timesyn [848] get 1-7
systemd-timesyn [848] set 1-32
```

I know `strace` could print more human friendly results:
```
setsockopt(7, SOL_NETLINK, NETLINK_EXT_ACK, [1], 4) = 0
setsockopt(7, SOL_NETLINK, NETLINK_GET_STRICT_CHK, [1], 4) = 0
setsockopt(7, SOL_NETLINK, NETLINK_PKTINFO, [1], 4) = 0
setsockopt(7, SOL_NETLINK, NETLINK_ADD_MEMBERSHIP, [1], 4) = 0
setsockopt(7, SOL_NETLINK, NETLINK_ADD_MEMBERSHIP, [5], 4) = 0
setsockopt(7, SOL_NETLINK, NETLINK_ADD_MEMBERSHIP, [9], 4) = 0
```

Are there any ways to get similar output with `bpftrace`?

**Comment by danobi** [comment]
You would have to write some code in bpftrace to do the translation. In the future, we may have support for enums: https://github.com/bpftrace/bpftrace/issues/599

But for all syscalls args, maybe. We'd need to think that through. I could potentially see value

**Reply by ajor** [reply]
Enum support is definitely on the todo list. It looks like the constants for these system calls are "#define" macros rather than enums though, and they don't appear in the BTF at all 😞:
https://github.com/torvalds/linux/blob/4cece764965020c22cff7665b18a012006359095/include/uapi/linux/netlink.h#L175

Strace appears to maintain its own mappings manually, which I don't really fancy doing: https://github.com/strace/strace/tree/master/src/xlat

The ideal solution would be to have the kernel's build process also generate BTF entries for these "#define"s somehow.

**Reply by jordalgo** [reply]
As an example for the bpftrace code you'd need - basically just a simple map from code to enum string e.g.
```
BEGIN {
        @errcodes[1] = "EPERM";
        @errcodes[2] = "ENOENT";
        @errcodes[3] = "ESRCH";
        @errcodes[4] = "EINTR";
        @errcodes[5] = "EIO";
        @errcodes[6] = "ENXIO";
        @errcodes[7] = "E2BIG";
        @errcodes[8] = "ENOEXEC";
 }
 ```
